### PR TITLE
Add expired status to task and expose task in-queue TTL to sdk

### PR DESCRIPTION
**Pull request by jsun-m**
Task dispatcher already automatically removes task from the list of pending tasks if it is pending up for 2 hours. However, the `STATUS` shown is the exact same status as when a `RUNNING` task times out (`TIMEOUT`). This results in a state that is confusing to the user as well developers trying to debug issues related to this task activity.

### CODE REVIEW

**Review by dleviminzi — APPROVED** [review]

**Comment by mernit** [comment]
Nice, this is going to be helpful

File: pkg/abstractions/function/function.go [hunk 1]
```diff
@@ -123,8 +123,23 @@ func (fs *RunCFunctionService) FunctionInvoke(in *pb.FunctionInvokeRequest, stre
 }
 
 func (fs *RunCFunctionService) invoke(ctx context.Context, authInfo *auth.AuthInfo, stubId string, payload *types.TaskPayload) (types.TaskInterface, error) {
-	policy := types.DefaultTaskPolicy
-	policy.Expires = time.Now().Add(time.Duration(functionDefaultTaskExpiration) * time.Second)
+	stub, err := fs.backendRepo.GetStubByExternalId(ctx, stubId)
```
Comment (luke-lombardi) [inline pkg/abstractions/function/function.go:126]: Kind of a shame we have to query for this here on every function invocation -- aren't we also call this a level up?

```func (fs *RunCFunctionService) FunctionInvoke(in *pb.FunctionInvokeRequest, stream pb.FunctionService_FunctionInvokeServer) error {
	authInfo, _ := auth.AuthInfoFromContext(stream.Context())
	ctx := stream.Context()

	task, err := fs.invoke(ctx, authInfo, in.StubId, &types.TaskPayload{
		Args: []interface{}{in.Args},
	})
	if err != nil {
		return err
	}

	go func() {
		stub, err := fs.backendRepo.GetStubByExternalId(ctx, in.StubId)
		if err != nil {
			log.Printf("error getting stub: %v", err)
			return
		}

		go fs.eventRepo.PushRunStubEvent(authInfo.Workspace.ExternalId, &stub.Stub)
	}()

	return fs.stream(ctx, stream, authInfo, task)
}```

File: pkg/gateway/services/stub.go [hunk 2]
```diff
@@ -32,21 +33,35 @@ func (gws *GatewayService) GetOrCreateStub(ctx context.Context, in *pb.GetOrCrea
 		autoscaler.TasksPerContainer = uint(in.Autoscaler.TasksPerContainer)
 	}
 
+	taskPolicy := types.TaskPolicy{
+		MaxRetries: uint(in.Retries),
+		Timeout:    int(in.Timeout),
+	}
+
+	if in.TaskPolicy.Expiration > 0 {
```
Comment (luke-lombardi) [inline pkg/gateway/services/stub.go:41]: Can you clean up this casting? Feels like a lot of nested stuff

File: pkg/gateway/services/stub.go [hunk 3]
```diff
@@ -32,21 +33,35 @@ func (gws *GatewayService) GetOrCreateStub(ctx context.Context, in *pb.GetOrCrea
...
+		taskPolicy.Expiration = uint32(math.Min(float64(in.TaskPolicy.Expiration), float64(types.MaxTaskExpirationS)))
+	}
+
+	if in.TaskPolicy.MaxRetries > 0 {
```
Comment (luke-lombardi) [inline pkg/gateway/services/stub.go:45]: same comment on these two

File: pkg/types/task.go [hunk 4]
```diff
@@ -108,10 +108,14 @@ var DefaultTaskPolicy = TaskPolicy{
 	Timeout:    3600,
 }
 
+var MaxTaskExpirationS = 24 * 60 * 60
+var MaxTaskRetries = 5
+
 type TaskPolicy struct {
 	MaxRetries uint      `json:"max_retries" redis:"max_retries"`
 	Timeout    int       `json:"timeout" redis:"timeout"`
 	Expires    time.Time `json:"expires" redis:"expires"`
```
Comment (luke-lombardi) [inline pkg/types/task.go:117]: I think it's confusing that we're adding an additional field here on task policy, expires and expiration sound the same to me. I know that they represent different things but on the go side I wouldn't know that they did

**Review by luke-lombardi — CHANGES_REQUESTED** [review]

File: pkg/gateway/services/stub.go (hunk 3 above)
Comment (jsun-m) [reply pkg/gateway/services/stub.go]: reorganized

File: pkg/gateway/services/stub.go (hunk 2 above)
Comment (jsun-m) [reply pkg/gateway/services/stub.go]: reorganized

File: pkg/abstractions/endpoint/buffer.go [hunk 5]
```diff
@@ -349,9 +353,16 @@ func (rb *RequestBuffer) handleHttpRequest(req request) {
 
 	resp, err := httpClient.Do(httpReq)
 	if err != nil {
-		req.ctx.JSON(http.StatusInternalServerError, map[string]interface{}{
-			"error": "Internal server error",
-		})
+		var netErr net.Error
+		if errors.As(err, &netErr) && netErr.Timeout() {
```
Comment (luke-lombardi) [inline pkg/abstractions/endpoint/buffer.go:357]: this second condition could be cleaned up

File: pkg/abstractions/function/function.go [hunk 6]
```diff
@@ -114,28 +115,30 @@ func (fs *RunCFunctionService) FunctionInvoke(in *pb.FunctionInvokeRequest, stre
...
+
+	policy := config.TaskPolicy
+	policy.Expires = time.Now().Add(time.Duration(policy.TTL) * time.Second)
 
 	task, err := fs.taskDispatcher.SendAndExecute(ctx, string(types.ExecutorFunction), authInfo, stubId, payload, policy)
 	if err != nil {
 		return nil, err
 	}
 
+	go fs.eventRepo.PushRunStubEvent(authInfo.Workspace.ExternalId, &stub.Stub)
```
Comment (luke-lombardi) [inline pkg/abstractions/function/function.go:141]: 👍

File: pkg/gateway/services/stub.go [hunk 7]
```diff
@@ -146,3 +147,29 @@ func (gws *GatewayService) DeployStub(ctx context.Context, in *pb.DeployStubRequ
...
 }
+
+func (gws *GatewayService) configureTaskPolicy(policy *pb.TaskPolicy, stubType types.StubType) types.TaskPolicy {
+	p := types.TaskPolicy{
+		MaxRetries: uint(math.Min(float64(policy.MaxRetries), float64(types.MaxTaskRetries))),
+		Timeout:    int(policy.Timeout),
+		TTL:        uint32(math.Min(float64(policy.Ttl), float64(types.MaxTaskTTL))),
+	}
+
+	if stubType.Kind() == types.StubTypeEndpoint {
```
Comment (luke-lombardi) [inline pkg/gateway/services/stub.go:158]: can this be a switch?

**Review by luke-lombardi — CHANGES_REQUESTED** [review]
left a couple more comments, getting close though 👍

File: pkg/abstractions/endpoint/buffer.go (hunk 5 above)
Comment (jsun-m) [reply pkg/abstractions/endpoint/buffer.go]: cleaned up

File: pkg/task/dispatch.go [hunk 8]
```diff
@@ -166,7 +166,7 @@ func (d *Dispatcher) monitor(ctx context.Context) {
 					continue
 				}
 
-				if !heartbeat && claimed {
```
Comment (jsun-m) [inline pkg/task/dispatch.go:169]: this claimed check is redundant cause if its not claimed it skips

**Review by luke-lombardi — APPROVED** [review]
👍

**Pull request merged**